### PR TITLE
Logging framework

### PR DESCRIPTION
**Pull request by MartinBoge**
Seamless logging and metric tracking with SPETLR. Introducing the `LogOrchestrator` and `LogTransformer` classes, these provide a plug-and-play solution for capturing valuable insights during your ETL workflows. Measure quantitative and qualitative values, count rows, track nulls, or implement custom metrics.

The `LogOrchestrator` extends the functionality of standard OETL orchestration, enabling automatic logging with `LogTransformer`. Add logging steps using `.log_with()` and let the `LogOrchestrator` handle the rest. Logging results are automatically written to handles provided at instantiation.

The `LogTransformer` is a base class for effortlessly creating transformers that log. The PR already provides a `CountLogTransformer` and a `NullLogTransformer`.

For reviewing I advise starting with the documentation and then look at each commit in order.

### CODE REVIEW

File: src/spetlr/etl/log/log_orchestrator.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        if len(self.log_transformers_output_keys) != 0:
+            # if only a single log transformer has been added its dataset output key is
+            # stored in a variable. This can then be written to the handles provided
+            # when the class was instantiated.
+            if len(self.log_transformers_output_keys) == 1:
+                dataset_key_to_log = self.log_transformers_output_keys
+
+            # if more than one log transformers has been added then these need to be
+            # unioned before writing them to the handles provided
+            if len(self.log_transformers_output_keys) > 1:
```
[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_orchestrator.py:90]: Could you elaborate on this?

File: src/spetlr/etl/log/log_transformer.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    # This is the default method for log many, if the logging in the subclass can
+    # be applied to each dataframe in the dataset separately. However, if the logging
+    # logic should be performed across multiple dataframes in the dataset, then the
+    # user should implement their own method.
+    @abstractmethod
+    def log_many(self, dataset: dataset_group) -> DataFrame:
+        for i, (key, df) in enumerate(dataset.items()):
+            if i == 0:
+                df_log = self.log(df).withColumn("DatasetInputKey", F.lit(key))
+            elif i >= 1:
```
[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_transformer.py:84]: Just to be explicit about how `i` should be viewed, but else would work just fine as well

File: src/spetlr/etl/log/log_orchestrator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LauJohansson) [reply src/spetlr/etl/log/log_orchestrator.py:90]: Yes.
The orchestrator uses UnionTransformerNC to uniion dataframes produces by  log transformers.

But, the .log() method returns a arbitraty dataframe depending on what the data engineer has defined.

```python

class MyFirstLog(LogTransformer):

   def log(self, df: DataFrame) -> DataFrame:
       return spark.dataframe("A string, B boolean")



class MySecondLog(LogTransformer):

   def log(self, df: DataFrame) -> DataFrame:
       return spark.dataframe("A string, B timestamp")

```

If I use log_with(MyFirstLog) and log_with(MySecondLog), these provided two different schemas. Wouldnt this make errors in the LogOrchestrator flow when trying to union the two log dataframes?

File: src/spetlr/etl/log/log_transformer.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        df = (
+            df_log.withColumn("LogName", F.lit(self.log_name).cast(StringType()))
+            .withColumn("LogTimestamp", F.lit(self.timestamp_now).cast(TimestampType()))
+            .withColumn("LogMethodName", F.lit(self.log_method_name).cast(StringType()))
+        )
+
+        # this creates an id per log transformer
+        df = df.withColumn("LogId", F.lit(str(uuid.uuid4())))
+
+        columns_final = [
```
[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_transformer.py:101]: Could just throw a warning?

File: src/spetlr/etl/log/log_transformer.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    # logic should be performed across multiple dataframes in the dataset, then the
+    # user should implement their own method.
+    @abstractmethod
+    def log_many(self, dataset: dataset_group) -> DataFrame:
+        for i, (key, df) in enumerate(dataset.items()):
+            if i == 0:
+                df_log = self.log(df).withColumn("DatasetInputKey", F.lit(key))
+            elif i >= 1:
+                df = self.log(df).withColumn("DatasetInputKey", F.lit(key))
+                df_log = df_log.union(df)
```
[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_transformer.py:86]: I don't follow? Is this related to line 83-86?

[the remark this answers is not in the document]
Comment (LauJohansson) [reply src/spetlr/etl/log/log_transformer.py:87]: Line 86 uses `.union` - are you sure you dont want to use `.unionByName` in stead?

[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_transformer.py:87]: For deduplication purposes?

[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_transformer.py:87]: I mean, yes that would be ideal if the user has identical dataframes in the input dataset

[the remark this answers is not in the document]
Comment (MartinBoge) [reply src/spetlr/etl/log/log_transformer.py:87]: fixed

**Review by LauJohansson — APPROVED** [review]

**Pull request merged**